src/main/scala/scalaz/stream/async/mutable/Queue.scala

```diff
@@ -158,24 +172,42 @@ private[stream] object Queue {
     }
 
     //dequeue one element from the queue
-    def dequeueOne(ref: ConsumerRef, cb: (Throwable \/ A => Unit)): Unit = {
-      queued.headOption match {
-        case Some(a) =>
-          S(cb(\/-(a)))
-          queued = queued.tail
-          signalSize(queued.size)
-          if (unAcked.size > 0 && bound > 0 && queued.size < bound) {
-            val ackCount = bound - queued.size min unAcked.size
-            unAcked.take(ackCount).foreach(cb => S(cb(\/-(()))))
-            unAcked = unAcked.drop(ackCount)
-          }
+    def dequeueBatch(ref: ConsumerRef, limit: Int, cb: (Throwable \/ Seq[A]) => Unit): Unit = {
+      if (queued.isEmpty) {
+        val cb2: Throwable \/ A => Unit = {
+          case l @ -\/(_) => cb(l)
+          case \/-(a) => cb(\/-(a :: Nil))
+        }
 
-        case None =>
-          val entry : (ConsumerRef, Throwable \/ A => Unit)  = (ref -> cb)
-          consumers = consumers :+ entry
+        val entry: (ConsumerRef, Throwable \/ A => Unit)  = (ref -> cb2)
+        consumers = consumers :+ entry
+      } else {
+        val (send, remainder) = if (limit <= 0)
+          (queued, Vector.empty[A])
+        else
+          queued splitAt limit
+
+        S { cb(\/-(send)) }
+
+        queued = remainder
+        signalSize(queued.size)
+        if (unAcked.size > 0 && bound > 0 && queued.size < bound) {
+          val ackCount = bound - queued.size min unAcked.size
+          unAcked.take(ackCount).foreach(cb => S(cb(\/-(()))))
+          unAcked = unAcked.drop(ackCount)
+        }
       }
     }
 
+    def dequeueOne(ref: ConsumerRef, cb: (Throwable \/ A => Unit)): Unit = {
```

Comment:
Can do.  I originally thought I was going to implement the `Dequeue` case against this method (as it had been) and then have a separate `DequeueBatch` actor case, but this just worked out more cleanly.